### ISSUE TITLE
pytorch prefers to install pytorch-cpu despite GPU availability

### ISSUE THREAD

**Issue by ashnair1** [conda-forge/pytorch-cpu-feedstock#102]
### Comment:

In `environment.yml`, if we specify conda-forge `pytorch` as a dependency, how does conda select between `pytorch-cpu` and `pytorch-gpu`?  Currently there seems to be 3 pytorch repos in conda-forge i.e. `pytorch`, `pytorch-cpu` and `pytorch-gpu` but only one feedstock i.e. this one.

**Comment by hmaarrfk** [comment]
It will try to install pytorch with cuda support if it can, otherwise it will install it will CPU support only.

The capabilities of the machine running conda/mamba install are evaluated.

**Comment by ashnair1** [comment]
Installing pytorch from conda-forge on Ubuntu 20.04 currently installs pytorch-cpu even though I have a gpu

```
conda create -n torchtest python=3.7
conda install -c conda-forge pytorch
```

yields

```
## Package Plan ##

  environment location: /home/ashwin/anaconda3/envs/torchtest

  added / updated specs:
    - pytorch


The following packages will be downloaded:

    package                    |            build
    ---------------------------|-----------------
    cffi-1.15.0                |   py37h036bc23_0         225 KB  conda-forge
    numpy-1.21.6               |   py37h976b520_0         6.1 MB  conda-forge
    pytorch-1.11.0             |cpu_py37h14e09b7_1        49.7 MB  conda-forge
    ------------------------------------------------------------
                                           Total:        56.0 MB

The following NEW packages will be INSTALLED:

  cffi               conda-forge/linux-64::cffi-1.15.0-py37h036bc23_0
  libblas            conda-forge/linux-64::libblas-3.9.0-14_linux64_mkl
  libcblas           conda-forge/linux-64::libcblas-3.9.0-14_linux64_mkl
  liblapack          conda-forge/linux-64::liblapack-3.9.0-14_linux64_mkl
  libprotobuf        conda-forge/linux-64::libprotobuf-3.20.1-h6239696_0
  llvm-openmp        conda-forge/linux-64::llvm-openmp-14.0.3-he0ac6c6_0
  mkl                conda-forge/linux-64::mkl-2022.0.1-h8d4b97c_803
  ninja              conda-forge/linux-64::ninja-1.10.2-h4bd325d_1
  numpy              conda-forge/linux-64::numpy-1.21.6-py37h976b520_0
  pycparser          conda-forge/noarch::pycparser-2.21-pyhd8ed1ab_0
  pytorch            conda-forge/linux-64::pytorch-1.11.0-cpu_py37h14e09b7_1
  sleef              conda-forge/linux-64::sleef-3.5.1-h9b69904_2
  tbb                conda-forge/linux-64::tbb-2021.5.0-h924138e_1
  typing_extensions  conda-forge/noarch::typing_extensions-4.2.0-pyha770c72_1

The following packages will be DOWNGRADED:

  _openmp_mutex                                   4.5-2_gnu --> 4.5-2_kmp_llvm


Proceed ([y]/n)? n
```

Do you know how the machine capabilities are evaluated?

**Comment by ashnair1** [comment]
It would be helpful to know how conda evaluates the machine capability. I know I can circumvent this issue by explicitly specifying `pytorch-gpu` via `conda create -n torchtest -c conda-forge python=3.7 pytorch-gpu` but that would defeat the purpose of having a pytorch package in conda-forge in the first place.

**Comment by h-vetinari** [comment]
Please post the output of `conda info` (as the issue template asks you to). One relevant bit of information there will be what's detected for the `__cuda` virtual package.

Conda-forge prefers the package without suffixes. The existence of `pytorch-gpu` & `pytorch-cpu` is for compatibility (at the time; AFAIK not relevant anymore) with the pytorch-channel, which offered these packages (to avoid spurious co-installation of several pytorches).

**Comment by ashnair1** [comment]
So according to this, conda is looking for packages with `__cuda=11.6=0`. I'm guessing since conda-forge pytorch only has binaries compiled with 11.1 and 11.2, conda is opting for cpu version.

**Comment by hmaarrfk** [comment]
I think there is also something about trying to minimize dependencies. I wonder if that is going on.

For what its worth, I can recreate what you showed, and the same behavior happens with conda and mamba.

**Comment by adamjstewart** [comment]
Are there any updates on this issue? A lot of libraries use this feedstock for their pytorch dependencies and we definitely don't want to be installing the CPU version if the GPU version would work on a system.

**Comment by isuruf** [comment]
`conda install pytorch-gpu` will force the installation of GPU version.

**Comment by adamjstewart** [comment]
Yes, but this doesn't work if you develop a software library and you want to be able to distribute a single `environment.yml` for all users (both users with and without a GPU). `conda install pytorch` should do the right thing.

**Comment by isuruf** [comment]
Please post the `environment.yml` and the output.

**Comment by isuruf** [comment]
and also `conda info`

**Comment by adamjstewart** [comment]
`environment.yml` can be seen [here](https://github.com/microsoft/torchgeo/blob/v0.2.1/environment.yml), `conda info` can be seen [here](https://github.com/conda-forge/pytorch-cpu-feedstock/issues/102#issuecomment-1117233372). Note that we recently switched to the pytorch channel instead of the conda-forge channel. @ashnair1 can you check if the problem also exists on the pytorch channel? I don't actually use conda lol.

**Comment by isuruf** [comment]
This issue is a consequence of https://github.com/conda-forge/pytorch-cpu-feedstock/issues/71

**Comment by adamjstewart** [comment]
Damn, it's crazy to see the hoops you have to jump through in the conda world to build different variants of the same package. I use [Spack](https://spack.io) and the spec model is much more powerful so you can say things like:
```console
$ spack install py-torch@1.11.0 +cuda cuda_arch=60 ^cuda@11.7.0 %gcc@12.1.0
$ spack install py-torch@master +rocm -nnpack -qnnpack -xnnpack
```
and build specific versions of PyTorch with or without certain features, certain versions of dependencies, and certain compilers. I think the most control you have in conda is:
```console
$ conda install pytorch-gpu=1.11.0
```
and hope that that's all you care about.

**Comment by isuruf** [comment]
Let's not rehash the binary distro vs source distro debate here.

**Comment by ashnair1** [comment]
@adamjstewart 

FWIW, this doesn't seem to be an issue with the pytorch channel. #71 explains the reasoning well enough.

I did notice one issue but I'm guessing it's a temporary one. 

If you currently try to run `conda install pytorch cudatoolkit -c pytorch` it defaults to installing a cpu version of pytorch. I assume this is because there is no torch built with the latest cudatoolkit i.e. 11.7. This can be checked here - https://download.pytorch.org/whl/torch_stable.html 

Running `conda install pytorch cudatoolkit=11.3 -c pytorch` or `conda install pytorch cudatoolkit=11.5 -c pytorch` works as expected

Basically as long as your pytorch and cudatoolkit versions match the whl distributed by the team it should be fine.

**Comment by isuruf** [comment]
@adamjstewart, @ashnair1, you could just use `pytorch-gpu` which works even when there's no GPU. (just adds a few dependencies)

**Comment by adamjstewart** [comment]
Honestly that might be a good workaround, although so is using the pytorch channel instead of the conda-forge channel. If that's the suggested approach, why not just get rid of `pytorch-cpu` and `pytorch-gpu` and only ship a single GPU-capable version on all systems? From our perspective, this is no longer an issue, but from your perspective if you want to keep the current system then it seems like someone will need to tackle #71. Maybe this issue can be closed as a duplicate of #71?

**Comment by hmaarrfk** [comment]
I think #71 explain why we kept the old names are kept around.

A PR would be welcome to fix #71. I think I simply forgot to include it the last time around.

I'm going to close this as a symptom of #71 to keep the discussion focused. However, I do care about ensuring this is good for users. It is just hard for us to test all usecases.

We are also quite interested in keeping some kind of compatibility upstream. See https://github.com/conda-forge/pytorch-cpu-feedstock/issues/74 However while they did once reach out to help package, they did not reach out during their package name change. Its just hard to cooperate with competing interests at mind.

**Comment by ngam** [comment]
> only ship a single GPU-capable version on all systems?

@adamjstewart we had a heated debate about this a while ago, but the reasoning is that some maintainers in this ecosystem prefer to ship as optimized a piece of software as possible without overloading, but more importantly, installing a cuda-fied version of pytorch/tensorflow will be on the order of 10x in terms of storage and bandwidth, hence unnecessarily wasteful. This explains it a little more: https://conda-forge.org/docs/user/tipsandtricks.html#installing-cuda-enabled-packages-like-tensorflow-and-pytorch

I was personally in favor of the ship-all-you-can-in-one approach, precisely because of problems like here (and to improve portability, etc.), but long-time maintainers around here do have a pretty good point... so hopefully we will find a smoother solution in the near future around this tailored approach :)

**Comment by hmaarrfk** [comment]
I'm going to reopen this issue so that others can find it. but I hope to have new builds out by next week.
